zerver/lib/zulip_update_announcements.py

```diff
@@ -0,0 +1,207 @@
+        realm=realm,
+        event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,
+        extra_data__contains={
+            RealmAuditLog.OLD_VALUE: None,
+            RealmAuditLog.NEW_VALUE: 1,
+            "property": "zulip_update_announcements_level",
+        },
+    ).first()
+    assert level_none_to_one_auditlog is not None
+    group_direct_message_sent_on = level_none_to_one_auditlog.event_time
+    return timezone_now() - group_direct_message_sent_on > timedelta(days=7)
+
+
+def internal_prep_zulip_update_announcements_stream_messages(
+    current_level: int, latest_level: int, sender: UserProfile, realm: Realm
+) -> List[Optional[SendMessageRequest]]:
+    message_requests = []
+    stream = realm.zulip_update_announcements_stream
+    assert stream is not None
+    with override_language(realm.default_language):
+        topic_name = str(realm.ZULIP_UPDATE_ANNOUNCEMENTS_TOPIC_NAME)
+        while current_level < latest_level:
+            content = _(get_zulip_update_announcements_message_for_level(level=current_level + 1))
+            message_requests.append(
+                internal_prep_stream_message(
+                    sender,
+                    stream,
+                    topic_name,
+                    content,
+                )
+            )
+            current_level += 1
+    return message_requests
+
+
+def send_messages_and_update_level(
+    realm: Realm,
+    new_zulip_update_announcements_level: int,
+    send_message_requests: List[Optional[SendMessageRequest]],
+) -> None:
+    sent_message_ids = []
+    if send_message_requests:
+        sent_messages = do_send_messages(send_message_requests)
+        sent_message_ids = [sent_message.message_id for sent_message in sent_messages]
+
+    RealmAuditLog.objects.create(
+        realm=realm,
+        event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,
+        event_time=timezone_now(),
+        extra_data={
+            RealmAuditLog.OLD_VALUE: realm.zulip_update_announcements_level,
+            RealmAuditLog.NEW_VALUE: new_zulip_update_announcements_level,
+            "property": "zulip_update_announcements_level",
+            "zulip_update_announcements_message_ids": sent_message_ids,
+        },
+    )
+
+    realm.zulip_update_announcements_level = new_zulip_update_announcements_level
+    realm.save(update_fields=["zulip_update_announcements_level"])
+
+
+def send_zulip_update_announcements() -> None:
+    latest_zulip_update_announcements_level = get_latest_zulip_update_announcements_level()
+
+    realms = get_realms_behind_zulip_update_announcements_level(
+        level=latest_zulip_update_announcements_level
+    )
+
+    for realm in realms:
+        realm_zulip_update_announcements_level = realm.zulip_update_announcements_level
+        sender = get_system_bot(settings.NOTIFICATION_BOT, realm.id)
+
+        messages = []
+        new_zulip_update_announcements_level = None
+        with transaction.atomic(savepoint=False):
```

Comment:
Oops, I realize I didn't say the thing thing -- it's inner the `transaction.atomic` in `do_send_messages` that needs this parameter, not this one. (Basically a transaction-inside-a-transaction makes a savepoint by default).

You can change it in a prep commit.

Reply:
Done.